purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt

```diff
@@ -0,0 +1,394 @@
+                    fetchPolicy = CacheFetchPolicy.FETCH_CURRENT,
+                    onError = {
+                        latch.countDown()
+                        fail("Expected success but got error: $it")
+                    },
+                    onSuccess = { receivedCustomerInfo ->
+                        assertCustomerInfoHasExpectedPurchaseData(receivedCustomerInfo)
+                        latch.countDown()
+                    }
+                )
+            }
+        }
+    }
+
+    @Test
+    fun doesNotEnterOfflineEntitlementsModeIfCachedCustomerInfoAndCustomerInfoRequestReturns500() {
+        ensureBlockFinishes { latch ->
+            setupTestWaitingForInitialRequests(
+                initialActivePurchases = initialActivePurchases
+            ) {
+                Purchases.sharedInstance.forceServerErrors = true
+
+                Purchases.sharedInstance.getCustomerInfoWith(
+                    fetchPolicy = CacheFetchPolicy.FETCH_CURRENT,
+                    onError = {
+                        Purchases.sharedInstance.getCustomerInfoWith(
+                            fetchPolicy = CacheFetchPolicy.CACHE_ONLY,
+                            onError = {
+                                latch.countDown()
+                                fail("Expected success but got error: $it")
+                            },
+                            onSuccess = {
+                                assertCustomerInfoDoesNotHavePurchaseData(it)
+                                latch.countDown()
+                            }
+                        )
+                    },
+                    onSuccess = {
+                        latch.countDown()
+                        fail("Expected error but got success: $it")
+                    }
+                )
+            }
+        }
+    }
+
+    @Test
+    fun entersOfflineEntitlementsModeIfPurchaseRequestReturns500() {
+        val storeProduct = StoreProductFactory.createGoogleStoreProduct()
+
+        ensureBlockFinishes { latch ->
+            setupTestWaitingForInitialRequests(
+                initialActivePurchases = emptyMap()
+            ) { activity ->
+                Purchases.sharedInstance.forceServerErrors = true
+
+                val receivedCustomerInfosInListener = mutableListOf<CustomerInfo>()
+                Purchases.sharedInstance.updatedCustomerInfoListener = UpdatedCustomerInfoListener {
+                    receivedCustomerInfosInListener.add(it)
+                }
+                mockPurchaseResult()
+                mockBillingAbstract.mockQueryProductDetails(queryProductDetailsSubsReturn = listOf(storeProduct))
+                Purchases.sharedInstance.purchaseWith(
+                    PurchaseParams.Builder(activity, storeProduct).build(),
+                    onError = { error, _ ->
+                        latch.countDown()
+                        fail("Expected success but got error: $error")
```

Comment:
Is it a problem that this is in this order? The code waiting for the latch would be notified before you fail the test.

Reply:
Right, I think we should completely remove the countdown in these error cases, I added it on an early iteration but it's not really needed, and can cause errors as you said. Good catch!